Framework/DataHandling/src/CompressEventAccumulator.cpp

```diff
@@ -0,0 +1,394 @@
+    m_findBin = EventList::findLogBin;
+    m_divisor = 1. / log1p(abs(divisor)); // use this to do change of base
+    m_offset = log(tof_min) * m_divisor;
+  } else {
+    throw std::runtime_error("Haven't implemented this compression binning strategy");
+  }
+}
+
+std::size_t CompressEventAccumulator::numberHistBins() const { return m_histogram_edges->size(); }
+
+template <typename INT_TYPE> double CompressEventAccumulator::getBinCenter(const INT_TYPE bin) const {
+  // work with iterators to reduce access time
+  const auto &binIter = m_histogram_edges->cbegin() + static_cast<std::vector<double>::difference_type>(bin);
+  return 0.5 * ((*binIter) + *(std::next(binIter)));
+}
+
+boost::optional<size_t> CompressEventAccumulator::findBin(const float tof) const {
+  // last parameter being false means don't find exact bin for raw events coming out of the file
+  return m_findBin(*m_histogram_edges.get(), static_cast<double>(tof), m_divisor, m_offset, false);
+}
+
+// ------------------------------------------------------------------------
+namespace { // anonymous
+
+constexpr size_t MAX_EVENTS_FOR_SPARSE_INT{20000};
+
+// minimum size of a vector to bother doing parallel_sort
+// the grain size shouldn't get too small or the overhead of threading/sync overwhelms everything else
+// TODO this can be tuned
+constexpr size_t MIN_VEC_LENGTH_PARALLEL_SORT{5000};
+
+/**
+ * Private class for implementation details of sparse event collection, when there are less events than bins
+ */
+class CompressSparseFloat : public CompressEventAccumulator {
+public:
+  // pass all arguments to the parent
+  CompressSparseFloat(std::shared_ptr<std::vector<double>> histogram_bin_edges, const size_t numEvents,
+                      const double divisor, CompressBinningMode bin_mode)
+      : CompressEventAccumulator(histogram_bin_edges, divisor, bin_mode), m_is_sorted(false) {
+    m_tof.reserve(numEvents);
+    m_initialized = true;
+  }
+
+  double totalWeight() const override { return static_cast<double>(m_tof.size()); }
+
+  /**
+   * This assumes that the event is within range of the fine histogram
+   */
+  void addEvent(const float tof) override {
+    // add events to the end of the list
+    m_tof.push_back(tof);
+  }
+
+private:
+  // this is marked because it is mostly used by createWeightedEvents and m_tof is mutable
+  void sort() const {
+    if (m_is_sorted)
+      return;
+
+    // empty events and singles are already sorted
+    if (m_tof.size() < 2) {
+      m_is_sorted = true;
+      return;
+    }
+
+    // calculate smallest tolerance
+    const auto &iter = m_histogram_edges->cbegin();
+    const auto delta = 1. / static_cast<float>((*(iter + 1)) - (*iter));
+
+    // parallel sort only if this is big enough of a vector
+    if (m_tof.size() < MIN_VEC_LENGTH_PARALLEL_SORT)
+      std::sort(m_tof.begin(), m_tof.end(), [delta](const auto &left, const auto &right) {
+        return std::floor(left * delta) < std::floor(right * delta);
```

Comment:
a hand-waiving calculation tells me that evaluating `left < right` is about three times faster than evaluating  `std::floor(left * delta) < std::floor(right * delta)`, so why use this?

Reply:
It is because the values get converted to `int` via `floor` rather than rounding for use as array indexes. I agree the individual compare is faster, but this doesn't need to move elements if the are "equal" within a single bin which is what this does.